contracts/root/withdrawManager/WithdrawManager.sol

```diff
@@ -47,142 +50,128 @@ contract WithdrawManager is WithdrawManagerStorage, IWithdrawManager {
-    // });
-
-    // // withdraw
-    // _withdraw(
-    //   _exitObject,
-
-    //   headerNumber,
-    //   withdrawBlockProof,
-
-    //   withdrawblockNumber,
-    //   withdrawblockTime,
-
-    //   withdrawblockTxRoot,
-    //   withdrawblockReceiptRoot,
-
-    //   path,
-    //   0
-    // );
+    PlasmaExit memory _exitObject = PlasmaExit({
+      owner: msg.sender,
+      token: rootToken,
+      receiptAmountOrNFTId: receiptAmountOrNFTId,
+      burnt: true
+    });
+
+    _withdraw(
+      _exitObject,
+      headerNumber,
+      withdrawBlockProof,
+      withdrawBlockNumber,
+      withdrawBlockTime,
+      withdrawBlockTxRoot,
+      withdrawBlockReceiptRoot,
+      path,
+      0 // oIndex
+    );
   }
 
-  function _processBurntReceipt(
-    bytes memory receiptBytes, bytes memory path, bytes memory receiptProof, bytes32 receiptRoot, address sender)
+  function _withdraw(
+    PlasmaExit memory exitObject,
+    uint256 headerNumber,
+    bytes memory withdrawBlockProof,
+    uint256 withdrawBlockNumber,
+    uint256 withdrawBlockTime,
+    bytes32 txRoot,
+    bytes32 receiptRoot,
+    bytes memory path,
+    uint8 oIndex)
     internal
-    view
-    returns (address rootToken, uint256 amountOrTokenId)
   {
-    RLPReader.RLPItem[] memory items = receiptBytes.toRlpItem().toList();
-    require(items.length == 4, "MALFORMED_RECEIPT");
-    // Do any other fields other than items[3] need to be checked?
-
-    // [3][1] -> [childTokenAddress, [WITHDRAW_EVENT_SIGNATURE, rootTokenAddress, sender], amount]
-    items = items[3].toList()[1].toList();
-    require(items.length == 3, "MALFORMED_RECEIPT"); // find a better msg
-    address childToken = items[0].toAddress();
-    amountOrTokenId = items[2].toUint();
-
-    // [3][1][1] -> [WITHDRAW_EVENT_SIGNATURE, rootTokenAddress, sender]
-    items = items[1].toList();
-    require(items.length == 3, "MALFORMED_RECEIPT"); // find a better msg;
+    uint256 startBlock;
+    bytes32 headerRoot;
+
+    // @todo writing a function to return just root and startBlock might save gas
+    (headerRoot, startBlock,,,) = rootChain.headerBlocks(headerNumber);
+
     require(
-      bytes32(items[0].toUint()) == WITHDRAW_EVENT_SIGNATURE,
-      "WITHDRAW_EVENT_SIGNATURE_NOT_FOUND"
+      keccak256(abi.encodePacked(withdrawBlockNumber, withdrawBlockTime, txRoot, receiptRoot))
+        .checkMembership(withdrawBlockNumber - startBlock, headerRoot, withdrawBlockProof),
+      "WITHDRAW_BLOCK_NOT_A_PART_OF_SUBMITTED_HEADER"
     );
 
-    rootToken = BytesLib.toAddress(items[1].toBytes(), 12);
-    require(
-      registry.rootToChildToken(rootToken) == childToken,
-      "INVALID_ROOT_TO_CHILD_TOKEN_MAPPING"
+    uint256 _exitId = (
+      headerNumber * HEADER_NUMBER_WEIGHT +
+      withdrawBlockNumber * WITHDRAW_BLOCK_NUMBER_WEIGHT +
+      path.toRlpItem().toBytes().toRlpItem().toUint() * 100000 +
+      oIndex
     );
 
-    // This check might be inconsequential. @todo check
```

Comment:
why would that be? we are extracting sender from tx Sig and double checking with param data as well.

Reply:
Same as our discussion that day about mandating whether validator/proposer himself should send the `submitHeaderBlock` tx. Similarly, If I burn my tokens on child chain, is it of any consequence that I myself should send the withdraw tx on rootChain?

Reply:
Correct. Removing the comment. FYI this code has moved to the lib.

Reply:
cool 👍.